Add task to download files from helix results container

safern opened the pull request:
Contributes to: #2076 

This pr adds a Task for helix jobs to be able to download a file from the results container after a helix run, i.e testResults.xml or coverage.xml. 

1. Every `HelixWorkItem` should include a metadata entry to specify which files it needs to download:
```xml
<HelixWorkItem Include="WorkItem">
   <DownloadFilesFromResults>testResults.xml</DownloadFilesFromResults>
</HelixWorkItem>
```
2. A destination directory in the host machine can be specified through an MSBuild property -- `$(HelixResultsDestinationDir)`.
    - If this property is not set and `BUILD_SOURCESDIRECTORY` property is available (predefined variable in Azure DevOps), it defaults to `$(BUILD_SOURCESDIRECTORY)/artifacts/helixresults`.
    - If `BUILD_SOURCESDIRECTORY` is empty, it would log a warning and will not execute the task.
3. A directory under this destination folder will be created per each Job created => each `HelixTargetQueue` for the same payload. This folder will be named with the JobId.
4. Under this Job directory, there will be a `metadata.txt` file, that will contain the following information:
```
HelixQueue=%(SentJob.HelixTargetQueue)
Configuration=$(HelixConfiguration)
Architecture=$(HelixArchitecture)
```
More metadata can be specified to be written into this file, through and ItemGroup => `HelixDownloadResultsMetadata`
5. Under the Job directory, there will be a directory for every `HelixWorkItem` that specified the `DownloadFilesFromResults` metadata, and under that folder it will contain the files that where available for download.
6. If a file is not found within the container, a Warning will be logged through MSBuild.

The structure of the directory will look as follows:
```
- HelixResultsDestinationDir \
  - JobId \
    - metadata.txt
    - HelixWorkItemName \
       - downloaded files.
```

cc: @ViktorHofer @danmosemsft @markwilkie @Chrisboh

jonfortescue commented on src/Microsoft.DotNet.Helix/JobSender/JobDefinition.cs:
```diff
@@ -148,12 +148,15 @@ public async Task<ISentJob> SendAsync(Action<string> log = null)
             {
-                resultsStorage = new ConnectionStringBlobHelper(ResultsStorageAccountConnectionString);
-                resultsStorageContainer = await resultsStorage.GetContainerAsync(TargetContainerName);
+                resultsStorageContainer = await storage.GetContainerAsync(TargetResultsContainerName);
```
The Helix API is able to handle not having a storage connection string specified and this should be the default behavior. This change rolls back the work done in #1991.

I think your work is fine, but we should support an unspecified connection string as well.

safern replied:
Ok, so when we don't specify a connection string, how do we get the storage container URI back so that we download the results? Is there a way to know that?

safern replied:
From reading your changes, I guess if there is no results container connection string, the container used as results container will be shared with the one used for artifacts and all that?

jonfortescue replied:
So in order to do what you want to do you will need to have the sdk create a container. Perhaps if you don't want to provide a connection string you can add an option to provide a results container name and have that trigger the creation of a container (a third option to that if/else block).

The current behavior is that the API creates a results container in the Helix Results storage account. This is not reported back to the user but does create some nice separation. @MattGal can explain more (I'm about to get on a plane). Also going to tag @alexperovich just in case.

safern replied:
Ok, talked to @alexperovich and I will condition it to use the connectionString when available, if not available and there are any workitems intenting to get a file from the results container, then we will use the `storage` account, until the Helix Api returns the results container as part of its response.

jonfortescue replied:
Okay, filed an issue for this work at https://github.com/dotnet/core-eng/issues/5324

jonfortescue replied:
Okay, filed an issue for this work at https://github.com/dotnet/core-eng/issues/5324; triaging it to GA for now

alexperovich commented on src/Microsoft.DotNet.Helix/Sdk/DownloadFromResultsContainer.cs:
```diff
@@ -0,0 +1,104 @@
+        public string ResultsContainer { get; set; }
+
+        [Required]
+        public string PathToDownload { get; set; }
```
This name feels weird. I think it should be OutputDirectory or something like that. The name needs to convey that the path needs to be a directory, and that this task writes to it.

safern replied:
Makes sense.

alexperovich commented on src/Microsoft.DotNet.Helix/Sdk/DownloadFromResultsContainer.cs:
```diff
@@ -0,0 +1,104 @@
+
+        private async Task DownloadFilesForWorkItem(ITaskItem workItem, string directoryPath)
+        {
+            if (workItem.GetRequiredMetadata(Log, "DownloadFilesFromResults", out string files))
```
This function logs an error if the metadata doesn't exist. I think this should just be a `TryGetMetadata` call so it doesn't fail the build if this metadata isn't set.

safern replied:
Makes sense.

alexperovich commented on src/Microsoft.DotNet.Helix/Sdk/DownloadFromResultsContainer.cs:
```diff
@@ -0,0 +1,104 @@
+                    {
+                        string destinationFile = Path.Combine(destinationDir.FullName, file);
+                        Log.LogMessage(MessageImportance.Normal, $"Downloading {file} => {destinationFile}...");
+                        CloudBlob blob = new CloudBlob(new Uri($"{ResultsContainer}{workItemName}/{file}"));
```
Is this intended to only work for anonymous jobs? This will need the SAS token if the job is not a public job.

alexperovich commented on src/Microsoft.DotNet.Helix/Sdk/DownloadFromResultsContainer.cs:
```diff
@@ -0,0 +1,104 @@
+
+        private void LogRequiredParameterError(string parameter)
+        {
+            Log.LogError($"Required parameter {nameof(parameter)} string was null or empty");
```
This should not be `nameof(parameter)` here. It should just be `parameter`.

safern replied:
Oops 😄

alexperovich commented on src/Microsoft.DotNet.Helix/Sdk/tools/download-results/DownloadFromResultsContainer.targets:
```diff
@@ -0,0 +1,30 @@
+  		  Condition="$(WaitForWorkItemCompletion)"
+  		  AfterTargets="CoreTest"
+  		  Inputs="unused"
+  		  Outputs="%(SentJob.HelixTargetQueue);%(SentJob.ResultsContainerUri)">
```
This should just batch on `%(SentJob.Identity)` the jobs will all have different identities.

safern replied:
Oh yeah you're right, good catch.

alexperovich commented on src/Microsoft.DotNet.Helix/Sdk/tools/download-results/DownloadFromResultsContainer.targets:
```diff
@@ -0,0 +1,30 @@
+
+    <PropertyGroup>
+      <ResultsDestinationPath Condition="'$(ResultsDestinationPath)' == '' AND '$(BUILD_SOURCESDIRECTORY)' != ''">$([MSBuild]::NormalizePath('$(BUILD_SOURCESDIRECTORY)', 'artifacts', helixresults'))</ResultsDestinationPath>
+      <ResultsDestinationPath Condition="'$(ResultsDestinationPath)' == ''">$([MSBuild]::NormalizePath('$(MSBuildStartupDirectory)', 'helixresults'))</ResultsDestinationPath>
```
This should skip the download if this is unspecified. It is a bit weird to download a bunch of stuff into the working directory auto-magically.

safern replied:
So you would say only set it to $(BUILD_SOURCESDIRECTORY) if available, if not, then skip the download?

alexperovich replied:
Yea. Dumping files into the source tree by default feels bad.

echesakovMSFT commented:
cc @RussKeldorph @jashook

safern commented:
@alexperovich @jonfortescue this is ready for another round of reviews.

alexperovich commented on src/Microsoft.DotNet.Helix/Sdk/tools/download-results/DownloadFromResultsContainer.targets:
```diff
@@ -0,0 +1,31 @@
+          Inputs="unused"
+          Outputs="%(SentJob.Identity)">
+    <ItemGroup>
+      <_workItemsWithDownloadMetadata Include="@(HelixWorkItem)" Condition="'%(HelixWorkItem.DownloadFilesFromResults)' != ''" />
```
I think you actually need to empty these item groups before adding things to them. IIRC this will currently write the metadata for jobs 1 and 2 together in the file for 2 because the item group has been filled with both.

safern replied:
I haven't seen this behavior locally, but it doesn't hurt to empty this item group.

Actually the metadata itemgroup was correctly written for multiple jobs. No duplication at all.

alexperovich replied:
Hrm, I wonder if there is some msbuild magic with batching that is making the items not show up.

safern replied:
Maybe because the ItemGroup is within the scope of the target and since we're batching it creates 2 different copies and evaluations of it? Basically 2 different scopes? @ericstj might know.

ericstj replied:
I believe the side-effects of a batched target aren't observable until after that target runs.

Try this:
```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp3.0</TargetFramework>
    <_newProperty>unset</_newProperty>
  </PropertyGroup>

  <ItemGroup>
    <TestItem Include="Abc" />
    <TestItem Include="123" />
  </ItemGroup>
  
  <Target Name="BatchTarget" Inputs="%(TestItem.Identity)" Outputs="Unused">
    <Message Text="BatchTarget Property: $(_newProperty)" Importance="High" />
    <ItemGroup>
      <_newItem Include="@(TestItem)" />
    </ItemGroup>
    <PropertyGroup>
      <_newProperty>@(TestItem)</_newProperty>
    </PropertyGroup>
    <Message Text="BatchTarget: @(_newItem)" Importance="High" />
  </Target>
  
  <Target Name="TestTarget" DependsOnTargets="BatchTarget">
    <Message Text="TestTarget Property: $(_newProperty)" Importance="High" />
    <Message Text="TestTarget: @(_newItem)" Importance="High" />
  </Target>

</Project>
```

Output of this is:
```
  BatchTarget Property: unset
  BatchTarget: Abc
  BatchTarget Property: unset
  BatchTarget: 123
  TestTarget Property: 123
  TestTarget: Abc;123
```

safern replied:
Since we don't intend to use this items outside the target I guess we're safe on leaving it as is.

ericstj commented on src/Microsoft.DotNet.Helix/Sdk/DownloadFromResultsContainer.cs:
```diff
@@ -0,0 +1,109 @@
+
+namespace Microsoft.DotNet.Helix.Sdk
+{
+    public class DownloadFromResultsContainer : BaseTask
```
Nit: this task should implement cancellation.

safern replied:
Done. Could you take a look please?

ericstj commented:
```
DownloadFromResultsContainer.cs(100,36): error CS1501: No overload for method 'DownloadToFileAsync' takes 3 arguments [D:\j\_work\1\s\src\Microsoft.DotNet.Helix\Sdk\Microsoft.DotNet.Helix.Sdk.csproj]
```

Hmm, do you need a newer Azure storage client? https://docs.microsoft.com/en-us/dotnet/api/microsoft.windowsazure.storage.blob.cloudblob.downloadtofileasync?view=azure-dotnet#Microsoft_WindowsAzure_Storage_Blob_CloudBlob_DownloadToFileAsync_System_String_System_IO_FileMode_System_Threading_CancellationToken_

safern commented:
> Hmm, do you need a newer Azure storage client?

Yes, they actually split up the new version into multiple packages. They stopped building WindowsAzure.Storage and created multiple packages. So figuring out which ones we need.

safern commented:
This is actually what I had to change to consume the new Azure SDK since they had some breaking changes and split the packages: https://github.com/dotnet/arcade/pull/2103/commits/49cb506ea8e8f27c73a8a70fccb01f1e5b30245c

FYI: @Anipik since I changed dependencies and namespaces on the CommitManager.

safern commented:
> This is actually what I had to change to consume the new Azure SDK since they had some breaking changes and split the packages: [49cb506](https://github.com/dotnet/arcade/commit/49cb506ea8e8f27c73a8a70fccb01f1e5b30245c)
> FYI: @Anipik since I changed dependencies and namespaces on the CommitManager.

Actually I just deleted the commit, Helix.SDK AzurePipelines task was not happy with the Newtonsoft upgrade required by the new azure SDKs. I will leave it as is, without flowing the cancellation token to `DowloadToFileAsync` since I don't have much time to invest on doing the upgrades, will revisit later to fix it if need it.

ericstj commented:
> Helix.SDK AzurePipelines task was not happy with the Newtonsoft upgrade required by the new azure SDKs

Yeah it probably needs the AssemblyResolver

I'm ok with you undoing the cancellation for now, but it's something you should consider moving forward for tasks which do long running things.  Not implementing cancellation means a control-C will hang until the task completes.  You can kill the process but that doesn't give a flushed log (so if you're working remote you might be missing the info for diagnosing).

safern commented:
I don't think it is the AssemblyResolver, since the AzureDevOpsTask has BaseTask as its base class and it actually enables the AssemblyResolver for the desktop implementation: https://github.com/dotnet/arcade/blob/master/src/Microsoft.DotNet.Helix/Sdk/BaseTask.Desktop.cs#L9

safern commented:
> I'm ok with you undoing the cancellation for now, but it's something you should consider moving forward for tasks which do long running things. Not implementing cancellation means a control-C will hang until the task completes. You can kill the process but that doesn't give a flushed log (so if you're working remote you might be missing the info for diagnosing).

Yeah, makes sense. Maybe moving the `ct.ThrowIfCancellationRequested()` inside the for loop so for every file it checks and it doesn't download all the files if a cancellation is requested before finishing?

alexperovich commented:
@safern @ericstj https://github.com/dotnet/arcade/issues/2130 tracks the work to make everything in the helix sdk cancelable. I created it a while back in dotnet/core-eng, but never had time to do it.
